doc/tutorials/space_shooter/step_4.md

```diff
@@ -0,0 +1,190 @@
+        amount: 4,
+        stepTime: .2,
+        textureSize: Vector2(8, 16),
+      ),
+    );
+
+    width = 25;
+    height = 50;
+    anchor = Anchor.center;
+  }
+}
+```
+
+So far, this does not introduces any new concept, we just created a component, setting
+up its animations attributes.
+
+A `Bullet` behavior is a simple one, it always moves towards to the top of the screen and should
+be removed from the game if they are not visible anymore, so let's add an `update` method to it
+and make it happen:
+
+```dart
+class Bullet extends SpriteAnimationComponent
+    with HasGameReference<SpaceShooterGame> {
+  Bullet({
+    super.position,
+  }) : super(size: Vector2(25, 50));
+
+  @override
+  Future<void> onLoad() async {
+    // Omitted
+  }
+
+  @override
+  void update(double dt) {
+    super.update(dt);
+
+    position.y += dt * -500;
+
+    if (position.y < -height) {
+      removeFromParent();
+    }
+  }
+}
+```
+
+The above code should be straight forward, but lets break it down:
+
+- We add to the bullet's y axis position at a rate of -500 pixels per second. Remember going up
+in the y axis means getting closer to `0` since the top left corner of the screen is `0, 0`.
+- If the y is smaller than the negative value of the bullet's height, means that the component is
+completely off the screen and it can be removed.
+
+Right, we have a `Bullet` class ready now, so lets make start to implement the action of shooting
+now. First thing, let's create two empty methods in the `Player` class, `startShooting` and
+`stopShooting`.
+
+```dart
+class Player extends SpriteAnimationComponent
+    with HasGameReference<SpaceShooterGame> {
+
+  // Rest of implementation omitted
+
+  void startShooting() {
+    // TODO
+  }
+
+  void stopShooting() {
+    // TODO
+  }
+}
+```
+
+And let's hook into those methods from the game class, we will do that by using the `onPanStart`
+and `onPanEnd` methods from the `PanDetector` mixin that we already have been using for the ship
+movement:
+
+```dart
+class SpaceShooterGame extends FlameGame with PanDetector {
+  late Player player;
+
+  // Rest of implementation omitted
+
+  @override
+  void onPanUpdate(DragUpdateInfo info) {
+    player.move(info.delta.global);
+  }
+
+  @override
+  void onPanStart(DragStartInfo info) {
+    player.startShooting();
+  }
+
+  @override
+  void onPanEnd(DragEndInfo info) {
+    player.stopShooting();
+  }
+}
+```
+
+We now have everything set up, so let's write the shooting routine in our player class.
+
+Remember, the shooting behavior will be adding bullets through time intervals when the player is
+dragging the starship.
+
+We could very much likely implement the time interval code manually, but Flame provides an component
+out of the box for that, the `TimerComponent`, so let's take advantage of it:
+
+
+```dart
+class Player extends SpriteAnimationComponent
+    with HasGameReference<SpaceShooterGame> {
+  late final TimerComponent _bulletSpawner;
```

Comment:
We should really add a flag to the `SpawnComponent` so that the position from the lambda can be used instead.

Reply:
Agree, maybe add an issue to flame to do that and we could revisit this tutorial when that happens?

Reply:
#2923